src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        public VSCommanding.CommandState GetCommandState(TypeCharCommandArgs args, Func<VSCommanding.CommandState> nextCommandHandler) => nextCommandHandler();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompleteStatementCommandHandler(ITextUndoHistoryRegistry undoHistoryRegistry, IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement_on_semicolon;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            BeforeExecuteCommand(args, executionContext);
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caretOpt = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretOpt.HasValue)
+            {
+                return;
+            }
+
+            var caret = caretOpt.Value;
+
+            var document = caret.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters since analysis 
+            // starting with a node containing delimiters assumes the caret is placed inside those delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentList(currentNode, caret, isCaretAtEndOfLine, syntaxFacts))
```

Comment:
this needs comments explaining why this is the right behavior.  i.e. i want to the *why* of it.  *why* is it the case that if we dont' look like a node in an argument list that we are bailing out.

Reply:
Comments added to the code.  This is where we opt in to the types of nodes that we are handling .  More can be added here as feature evolves.